library/x509write_csr.c

```diff
@@ -85,6 +85,87 @@ int mbedtls_x509write_csr_set_extension(mbedtls_x509write_csr *ctx,
                                       critical, val, val_len);
 }
 
+int mbedtls_x509write_csr_set_subject_alternative_name(mbedtls_x509write_csr *ctx,
+                                                       const mbedtls_x509_san_list *san_list)
+{
+    int ret = 0;
+    const mbedtls_x509_san_list *cur = san_list;
+    unsigned char *buf;
+    unsigned char *p;
+    size_t len;
+    size_t buflen = 0;
+
+    /* Determine the maximum size of the SubjectAltName list */
+    while (cur != NULL) {
+        if (cur->node.len <= 0) {
+            return 0;
+        }
+
+        /* Calculate size of the required buffer:
+         * + length of value for each name entry,
+         * + maximum 4 bytes for the length field,
+         * + 1 byte for the tag/type.
+         */
+        buflen += cur->node.len + 4 + 1;
+
+        cur = cur->next;
+    }
+
+    /* Add the extra length field and tag */
+    buflen += 4 + 1;
+
+    /* Allocate buffer */
+    buf = mbedtls_calloc(1, buflen);
```

Comment:
There's one problem with dynamic allocation here. `MBEDTLS_ASN1_CHK_ADD` will return if the length is exceeded, and the buffer will not be freed. I'm afraid that we'll probably need to stick to static allocation if the buffer is internal.

Reply:
See [here](https://github.com/Mbed-TLS/mbedtls/blob/development/library/x509write_crt.c#L328) for comparison.

Reply:
Yes, this is an issue. Need to investigate, but it seems static allocation can solve the problem here and should be sufficient.

Reply:
When I looked at the code I changed my mind. We can't use static buffer with fixed length as we can't assume the max size. Maybe it would be fine only for short stings that are at this stage, but moving forward when `directoryName` and `otherName` will be added using static buffer is too risky. Additionally we shouldn't allocate big buffers on stack.

I would go with checking status of each operation and `goto error` on failure and release memory.

Reply:
Hmm
Usually when a CSR is freed the underlying SAN buffers are not freed because they come raw from the input data buffer. However, in this case they were allocated by us, so when `mbedtls_x509_csr_free()` is called, they should be freed too. Some way to distinguish these cases and do the right thing in `mbedtls_x509_csr_free()` is needed.

Reply:
Ignore above, I had not properly understood the API.